client/client.go

```diff
@@ -1396,169 +1478,175 @@ func (c *Client) deriveToken(alloc *structs.Allocation, taskNames []string, vcli
-				port := strconv.FormatInt(int64(s.ServicePort), 10)
-				addr := s.ServiceAddress
-				if addr == "" {
-					addr = s.Address
-				}
-				serverAddr := net.JoinHostPort(addr, port)
-				serverEndpoint, err := rpcproxy.NewServerEndpoint(serverAddr)
+			// Successfully received the Server peers list of the correct
+			// region
+			for _, p := range peers {
+				addr, err := net.ResolveTCPAddr("tcp", p)
 				if err != nil {
 					mErr.Errors = append(mErr.Errors, err)
-					continue
-				}
-				var peers []string
-				if err := c.connPool.RPC(c.Region(), serverEndpoint.Addr, c.RPCMajorVersion(), "Status.Peers", nomadRPCArgs, &peers); err != nil {
-					mErr.Errors = append(mErr.Errors, err)
-					continue
-				}
-				// Successfully received the Server peers list of the correct
-				// region
-				if len(peers) != 0 {
-					nomadServerServices = append(nomadServerServices, peers...)
-					break
 				}
+				servers = append(servers, &endpoint{name: p, addr: addr})
 			}
-			// Break if at least one Nomad Server was successfully pinged
-			if len(nomadServerServices) > 0 {
-				break
+			if len(servers) > 0 {
+				break DISCOLOOP
 			}
 		}
-		if len(nomadServerServices) == 0 {
-			if len(mErr.Errors) > 0 {
-				return mErr.ErrorOrNil()
-			}
-
-			return fmt.Errorf("no Nomad Servers advertising service %q in Consul datacenters: %q", nomadServerServiceName, dcs)
+	}
+	if len(servers) == 0 {
+		if len(mErr.Errors) > 0 {
+			return mErr.ErrorOrNil()
 		}
+		return fmt.Errorf("no Nomad Servers advertising service %q in Consul datacenters: %+q", serviceName, dcs)
+	}
 
-		// Log the servers we are adding
-		c.logger.Printf("[INFO] client.consul: bootstrap adding following Servers: %q", nomadServerServices)
+	c.logger.Printf("[INFO] client.consul: discovered following Servers: %s", servers)
+	c.servers.set(servers)
 
-		c.heartbeatLock.Lock()
-		if atomic.LoadInt32(&c.lastHeartbeatFromQuorum) == 1 && now.Before(c.consulPullHeartbeatDeadline) {
-			c.heartbeatLock.Unlock()
-			// Common, healthy path
-			if err := c.rpcProxy.SetBackupServers(nomadServerServices); err != nil {
-				return fmt.Errorf("client.consul: unable to set backup servers: %v", err)
-			}
-		} else {
-			c.heartbeatLock.Unlock()
-			// If this Client is talking with a Server that
-			// doesn't have a leader, and we have exceeded the
-			// consulPullHeartbeatDeadline, change the call from
-			// SetBackupServers() to calling AddPrimaryServer()
-			// in order to allow the Clients to randomly begin
-			// considering all known Nomad servers and
-			// eventually, hopefully, find their way to a Nomad
-			// Server that has quorum (assuming Consul has a
-			// server list that is in the majority).
-			for _, s := range nomadServerServices {
-				c.rpcProxy.AddPrimaryServer(s)
+	// Notify waiting rpc calls. Wait briefly in case initial rpc
+	// just failed but the calling goroutine isn't selecting on
+	// discovered yet.
+	const dur = 50 * time.Millisecond
```

Comment:
Removed.

I think this was a clumsy workaround to get around startup races early on that AFAICT (both looking at the code again and testing it manually) is no longer necessary.